pkg/util/cert.go

```diff
@@ -104,3 +108,83 @@ func getKeyPair(pubKeyPath, privKeyPath string) (*rsa.PrivateKey, error) {
 
 	return rsaPrivKey, nil
 }
+
+func IsCertAllowed(clusterNetwork []string, serviceNetwork []string, certPath string, extraNames []string) (bool, error) {
+	certsSNIs, err := GetSNIsFromCert(certPath, extraNames)
+	if err != nil {
+		return false, err
+	}
+
+	// iterate over the SNIs and generate kubeconfig files
+	// TODO: do we want to read cert.IPAddresses?
+	for _, dns := range certsSNIs {
+		// check if SNI is allowed (non local or wildcard)
+		if !VerifyAllowedSNI(clusterNetwork, serviceNetwork, dns) {
+			klog.Infof("Certificate SNI is not allowed %s", dns)
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+// GetSNIsFromCert - get list of unique SNIs from certificate
+func GetSNIsFromCert(certPath string, extraNames []string) ([]string, error) {
+	pemByte, err := os.ReadFile(certPath)
+	if err != nil {
+		return nil, err
+	}
+	block, _ := pem.Decode(pemByte)
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		log.Fatal(err)
+	}
+	klog.Infof("CN contains %s", cert.Subject.CommonName)
+	klog.Infof("SAN contains %v DNSNames", cert.DNSNames)
+
+	// create one list to hold all the SNIs
+	certsSNIs := append(cert.DNSNames, cert.Subject.CommonName)
+
+	// add  Certificate SAN ipaddress
+	for _, ipaddress := range cert.IPAddresses {
+		certsSNIs = append(certsSNIs, ipaddress.String())
+	}
+	// add Configuration names
+	if len(extraNames) > 0 {
+		certsSNIs = append(certsSNIs, extraNames...)
+	}
+	return certsSNIs, nil
+}
+
+// IsWildcardDNS - check if DNS is a wildcard
+func IsWildcardDNS(val string) bool {
+	return strings.Contains(val, "*.")
+}
+
+func IsWildcardEntry(val string) bool {
+	return strings.Contains(val, "*.")
+}
+
+// verifyAllowedSNI checks if sni is allowed
+// return bool: true, false
+func VerifyAllowedSNI(clusterNetwork []string, serviceNetwork []string, sni string) bool {
+	forbiddenValues := []string{"localhost", "svc.cluster.local", "openshift.default.svc", "kubernetes.default.svc"}
+
+	ipAddress := tcpnet.ParseIP(sni)
+
+	//check if IPAddress or DNS record
+	if ipAddress == nil {
+		for _, val := range forbiddenValues {
+			if strings.Contains(sni, val) {
```

Comment:
So this will match SNI like `blalocalhostbla` but should it? What about `test.svc.cluster.local` - is this legal or not?

Reply:
for locahost you are correct we should check for exact match,
and   `test.svc.cluster.local` SNI will be rejected since its contained within forbidden value `svc.cluster.local`